haystack/components/tools/tool_invoker.py

```diff
@@ -609,80 +611,116 @@ async def run_async(
-                    tool_messages.append(ChatMessage.from_tool(tool_result=error_message, origin=tool_call, error=True))
-                    continue
-
-                # 3) Merge outputs into state
-                try:
-                    self._merge_tool_outputs(tool_to_invoke, tool_result, state)
-                except Exception as e:
-                    try:
+                    # Check if the tool is available, otherwise return an error message
+                    if tool_name not in self._tools_with_names:
                         error_message = self._handle_error(
-                            ToolOutputMergeError(f"Failed to merge tool outputs from tool {tool_name} into State: {e}")
+                            ToolNotFoundException(tool_name, list(self._tools_with_names.keys()))
                         )
                         tool_messages.append(
                             ChatMessage.from_tool(tool_result=error_message, origin=tool_call, error=True)
                         )
                         continue
-                    except ToolOutputMergeError as propagated_e:
-                        # Re-raise with proper error chain
-                        raise propagated_e from e
 
-                # 4) Prepare the tool result ChatMessage message
-                tool_messages.append(
-                    self._prepare_tool_result_message(
-                        result=tool_result, tool_call=tool_call, tool_to_invoke=tool_to_invoke
-                    )
-                )
+                    tool_to_invoke = self._tools_with_names[tool_name]
+
+                    # 1) Combine user + state inputs
+                    llm_args = tool_call.arguments.copy()
+                    final_args = self._inject_state_args(tool_to_invoke, llm_args, state)
+
+                    # Check whether to inject streaming_callback
+                    if (
+                        resolved_enable_streaming_passthrough
+                        and streaming_callback is not None
+                        and "streaming_callback" not in final_args
+                    ):
+                        invoke_params = self._get_func_params(tool_to_invoke)
+                        if "streaming_callback" in invoke_params:
+                            final_args["streaming_callback"] = streaming_callback
+
+                    # Dispatch each call into our local executor
+                    task = invoke_tool_safely(executor, tool_to_invoke, final_args)
+                    tool_call_tasks.append(task)
+                    valid_tool_calls.append((tool_call, tool_to_invoke))
+
+            # Execute all valid tool calls concurrently
+            if tool_call_tasks:
+                tool_results = await asyncio.gather(*tool_call_tasks)  # No return_exceptions since we handle in wrapper
+
+                # Process results
+                for (tool_call, tool_to_invoke), tool_result in zip(valid_tool_calls, tool_results):
```

Comment:
I think here it may be better to do `for i, ((tool_call, tool_to_invoke), tool_result) in enumerate(zip(valid_tool_calls, tool_results)):` (so process results in original order)...